WalletWasabi/Crypto/ZeroKnowledge/NonInteractive/FiatShamirTransform/Knowledge.cs

```diff
@@ -0,0 +1,78 @@
+using System.Collections.Generic;
+using System.Linq;
+using WalletWasabi.Crypto.Randomness;
+using WalletWasabi.Helpers;
+
+namespace WalletWasabi.Crypto.ZeroKnowledge.NonInteractive.FiatShamirTransform
+{
+	public class Knowledge : IKnowledge
+	{
+		// Although in principle this could be an interface type, since only one
+		// type of statement is needed for our proofs this is specialized to only
+		// that type.
+		public Knowledge(LinearRelation.Knowledge knowledge)
+		{
+			LinearRelation = Guard.NotNull(nameof(knowledge), knowledge);
```

Comment:
Is it a good idea to name something `knowledge` and then assign it to `LinearRelation`? Should the parameter be possibly named `linearRelation` too?

Reply:
probably not. since there are two classes called `Knowledge`, each at a different level of abstraction, containing each other, and the compiler told me i can't name a property the same as the class containing it (i guess namespace collision with constructors? although i thought they were syntactically distinct...)

because the leaf level is not polymorphic, i thought the next most logical thing is to name it after the namespace., so i guess this means the constructor argument should be renamed.

or maybe we can come up with a better way of naming this.

I since learned about qualified imports, so I think that importing the `LinearRelation` namespace under a different name can help.

Reply:
bikeshedded back to IProver and IVerifier, so this naming conflict is gone